File: program/OOversion/MCS.cpp
#include<bits/stdc++.h>
using namespace std;

class MCS {
public:
    int V;
    int E;
    vector<vector<int>> adjList;
    vector<bool> isVisited;
    vector<set<int>> MCSsets;
    vector<vector<int>> MCS_orderings;
    // contructor
    MCS(int V, int E, vector<pair<int, int>>& edges) {
        this->V = V;
        this->E = E;
        adjList.resize(V);
        isVisited.resize(V, false);
        MCSsets.resize(V, {});
        for (int i = 0; i < E; i++) {
            adjList[edges[i].first].push_back(edges[i].second);
            adjList[edges[i].second].push_back(edges[i].first);
        }
    }
    // checking whether the set associated with the vertex is maximal under set inclusion and has maximum cardinality (for MCS)
    bool isSetMaximalCardinality(int u) {
        int maxSize = 0;
        for (int i = 0; i < V; i++) {
            if (!isVisited[i]) {
                maxSize = max(maxSize, (int) MCSsets[i].size());
            } 
        }
        if (MCSsets[u].size() == maxSize) return true;
        else return false;
    }
    // MCS algorithm
    void performMCS(vector<int>& ordering) {
        if (ordering.size() == V) {
            MCS_orderings.push_back(ordering);
            return;
        }
        for (int i = 0; i < V; i++) {
            if (!isVisited[i]) {
                if (MCSsets[i].size() > 0 && isSetMaximalCardinality(i)) {
                    isVisited[i] = true;
                    ordering.push_back(i);
                    for (int neighbour : adjList[i]) {
                        if (!isVisited[neighbour]) {
                            MCSsets[neighbour].insert(ordering.size());
                        }
                    } 
                    // explore
                    performMCS(ordering);
                    // backtrack
                    for (int neighbour : adjList[i]) {
                        if (!isVisited[neighbour]) {
                            MCSsets[neighbour].erase(ordering.size());
                        }
                    }
                    isVisited[i] = false;
                    ordering.pop_back();
                } 
            }
        }    
    }
    void performAllPossible() {
        vector<int> ordering;
        for (int i = 0; i < V; i++) {
            MCSsets[i].insert(V + 1);
            performMCS(ordering);
            MCSsets[i].erase(V + 1);
        }
    }
    int getSize() {
        return this->MCS_orderings.size();
    }
};

File: generic_subgraph/mcs.py
class MCS:

    def __init__(self, v, e, edges) -> None: #constructor
        self.v = v #vertices
        self.e = e #edges
        self.adjList = [[] for i in range(v)]
        self.MCSsets = [set([]) for i in range(v)]
        self.MCS_orderings = set([])
        self.vis = [False for i in range(v)]
        for e in edges:
            self.adjList[e[0]].append(e[1])
            self.adjList[e[1]].append(e[0])
    
    def isSetMaximumCardinality(self, u) -> bool:
        mxSize = 0
        for node in range(self.v):
            if self.vis[node]==False:
                mxSize = max(mxSize, len(self.MCSsets[node]))
        return True if len(self.MCSsets[u])==mxSize else False
    
    def performMCS(self, currentOrdering) -> None:
        if len(currentOrdering)==self.v:
            self.MCS_orderings.add(tuple(currentOrdering))
            return
        for node in range(self.v):
            if self.vis[node]==False:
                if len(self.MCSsets[node])>0 and self.isSetMaximumCardinality(node):
                    self.vis[node]=True
                    currentOrdering.append(node)
                    for neighbour in self.adjList[node]:
                        if self.vis[neighbour]==False:
                            self.MCSsets[neighbour].add(len(currentOrdering))
                    self.performMCS(currentOrdering)
                    for neighbour in self.adjList[node]:
                        if self.vis[neighbour]==False:
                            self.MCSsets[neighbour].remove(len(currentOrdering))
                    self.vis[node]=False
                    currentOrdering.pop()

    def performAllPossible(self) -> None:
        currentOrdering = []
        for u in range(self.v):
            self.MCSsets[u].add(self.v+1)
            self.performMCS(currentOrdering)
            self.MCSsets[u].remove(self.v+1)

    def getSize(self):
        return len(self.MCS_orderings)